vignettes/articles/ui-cli-conversion.Rmd

```diff
@@ -0,0 +1,442 @@
+
+* `indent()` turns out to still be useful in `ui_code_snippet()`, so I've inlined
+  it there, to avoid any reliance on definitions in ui-legacy.R.
+* `ui_bullet()` has been renamed to `ui_legacy_bullet()` for auto-completion happiness with the new `ui_bullets()`.
+
+Let's cover `ui_silence()` while we're here, which *is* exported.
+It's just a `withr::with_*()` function for executing `code` with `usethis.quiet = TRUE`.
+
+```{r}
+#| eval: false
+ui_silence <- function(code) {
+  withr::with_options(list(usethis.quiet = TRUE), code)
+}
+```
+
+## Inline styles
+
+### Legacy functions
+
+usethis has its own inline styles (mostly) for use inside functions like `ui_todo()`:
+
+* `ui_field()`
+* `ui_value()`
+* `ui_path()`
+* `ui_code()`
+* `ui_unset()`
+
+```{asciicast}
+# why is this block truncated from the top in the rendered document?
+new_val <- "oxnard"
+x <- glue("{ui_field('name')} set to {ui_value(new_val)}")
+dput(x)
+ui_done(x)
+```
+
+The inline styles enact some combination of:
+
+* Color, e.g. `crayon::green(x)`
+* Collapsing a collection of things to one thing, e.g. `c("a", "b", "c")` to "a, b, c"
+* Quoting, e.g. `encodeString(x, quote = "'")`
+
+`ui_path()` is special because it potentially modifies the input before styling it.
+`ui_path()` first makes the path relative to a specific base (by default, the active project root) and, if the path is a directory, it also ensures there is a trailing `/`.
+
+`ui_unset()` is a special purpose helper used when we need to report that something is unknown, not configured, nonexistent, etc.
+
+```{asciicast}
+# why is this block truncated from the top in the rendered document?
+x <- glue("Your super secret password is {ui_unset()}.")
+dput(x)
+ui_info(x)
+```
+
+### cli replacements
+
+In general, we can move towards cli's native inline-markup:
+<https://cli.r-lib.org/reference/inline-markup.html>
+
+Here's the general conversion plan:
+
+-   `ui_field()` becomes `{.field blah}`. In `usethis_theme()`, I tweak the `.field` style to apply single quotes if color is not available, which is what `ui_field()` has always done.
+-   `ui_value()` becomes `{.val value}`.
+-   `ui_path()` is connected to `{.path path/to/thing}`, but, as explained above, `ui_path()` also does more. Therefore, I abstracted the "path math" into an internal helper, `ui_path_impl()`, which is aliased to `pth()` for compactness. Here's a typical conversion:
+    ```{r}
+    #| eval: false
+    # using legacy functions
+    ui_done("Setting {ui_field('LazyData')} to \\
+             {ui_value('true')} in {ui_path('DESCRIPTION')}")
+    # using new cli-based ui
+    ui_bullets(c(
+      "v" = "Setting {.field LazyData} to {.val true} in {.path {pth('DESCRIPTION')}}."
+    ))
+    ```
+    It would be nice to create a custom inline class, e.g. `{.ui_path {some_path}}`, which I have done in, e.g., googledrive.
+    But it's not easy to do this *while still inheriting cli's `file:` hyperlink behaviour*, which is very desirable.
```

Comment:
What if you used `format_inline()` + `.file` in your custom inline helper? Or do you think the usethis path logic is sufficiently useful that it should move to cli?

Reply:
I guess I'd rather not "explode" (as in pre-format) filepaths like this. It introduces a special behaviour that I suspect would come back to haunt me. In general, I try to keep the actual formatting as late as possible to reduce the chance of calling `format_inline()` under one set of conditions, but then having another set hold when the user actually sees the message.

Reply:
In terms of having something similar in cli, I'm not sure how you would pass the `base` information to an inline style? I don't see how to pass 2 pieces of info to an inline style (display path + optional base to make it relative to).